sdk/core/AzureCore/Source/DataStructures/ResumableOperationQueue.swift

```diff
@@ -0,0 +1,148 @@
+        }
+
+        set {
+            operationQueue.name = newValue
+        }
+    }
+
+    public var maxConcurrentOperationCount: Int {
+        get {
+            return operationQueue.maxConcurrentOperationCount
+        }
+
+        set {
+            operationQueue.maxConcurrentOperationCount = newValue
+        }
+    }
+
+    public var count: Int {
+        return operations.count
+    }
+
+    public var queueCount: Int {
+        return operationQueue.operationCount
+    }
+
+    public let removeOnCompletion: Bool
+
+    // MARK: Initializers
+
+    public init(name: String, delegate: ResumableOperationQueueDelegate? = nil, removeOnCompletion: Bool = false) {
+        self.removeOnCompletion = removeOnCompletion
+        self.name = name
+        self.delegate = delegate
+    }
+
+    // MARK: Public Methods
+
+    public func add(_ operation: ResumableOperation) {
+        add([operation])
+    }
+
+    public func add(_ operations: [ResumableOperation]) {
+        let states = operations.map { $0.state.rawValue }
+        for state in Set(states) {
+            let filteredOps = operations.filter { $0.state.rawValue == state }
+            var transferState = TransferState(rawValue: state) ?? .unknown
+            let allowed: [TransferState] = [.pending, .inProgress]
+            if allowed.contains(transferState) {
+                // reset inProgress back to pending since they may be scheduled differently
+                transferState = .pending
+
+                // For Pending or InProgress operations, need to requeue the operations
+                for operation in filteredOps {
+                    operation.delegate = operation.delegate ?? delegate
+                    let operationClosure = operation.completionBlock
+                    operation.completionBlock = {
+                        operationClosure?()
+                        if operation.isCancelled {
+                            return
+                        }
+                        operation.internalState = .complete
+                    }
+                    self.operations.append(operation)
+                }
+                operationQueue.addOperations(filteredOps, waitUntilFinished: false)
+            }
+        }
+    }
+
+    public func cancel(_ operation: ResumableOperation) {
+        operation.cancel()
+    }
+
+    public func clear() {
+        operationQueue.cancelAllOperations()
+        operations.removeAll()
+    }
+
+    internal func removeFromDataStore(_ operation: ResumableOperation) {
+        guard let index = operations.firstIndex(of: operation) else { return }
+        operations.remove(at: index)
+    }
+
+    public func pause(_ operation: ResumableOperation) {
+        operation.pause()
+        removeFromDataStore(operation)
+    }
+
+    public func remove(_ operation: ResumableOperation) {
+        removeFromDataStore(operation)
+
+        // Cancel the operation.
+        if !(operation as Operation).isFinished {
```

Comment:
Odd that this cast is needed here since `ResumableOperation` is a subclass of `Operation`.